### PR TITLE
update tensorflow.net to 0.20.0

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [x] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [x] You have included any necessary tests in the same PR.

# Related issue
#5401

### CODE REVIEW

**Comment by LittleLittleCloud** [comment]
Hi @Oceania2018, since there are some API changes in tensorflow.net, so I make a few changes in Microsoft.ML.Tensorflow to make usage of the new API. Would you be willing to review this PR?

Also @codemzs, as the creator of Microsoft.ML.Tensorflow, could you also review it as well.

**Review by Oceania2018 — APPROVED** [review]
https://github.com/dotnet/machinelearning/issues/5401

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs [hunk 1]
```diff
@@ -837,7 +838,7 @@ private Tensor CastDataAndReturnAsTensor(T[] data)
                         bytes[i] = Encoding.UTF8.GetBytes(((ReadOnlyMemory<char>)(object)data[i]).ToArray());
                     }
 
-                    return new Tensor(bytes, _tfShape.dims.Select(x => (long)x).ToArray());
+                    return new Tensor(new NDArray(bytes, _tfShape));
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:841]: @Oceania2018 
the type of bytes here is byte[][], Can you confirm it is the right way to create a 2-d byte tensor.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs [hunk 2]
```diff
@@ -752,15 +752,15 @@ private void CheckTrainingParameters(Options options)
...
+            var resizedImage = tf.image.resize_bilinear(decodedImage4d, resizeShapeAsInt, false, name: "ResizeTensor");
             return (jpegData, resizedImage);
         }
 
         private static Tensor EncodeByteAsString(VBuffer<byte> buffer)
         {
             int length = buffer.Length;
-            var size = c_api.TF_StringEncodedSize((UIntPtr)length);
-            var handle = c_api.TF_AllocateTensor(TF_DataType.TF_STRING, IntPtr.Zero, 0, (UIntPtr)((ulong)size + 8));
+            var size = c_api.TF_StringEncodedSize((ulong)length);
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:762]: @codemzs Can you confirm the change in `EncodeByteAsString` function is correct?

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs [hunk 3]
```diff
@@ -769,7 +769,7 @@ private static Tensor EncodeByteAsString(VBuffer<byte> buffer)
             unsafe
             {
                 fixed (byte* src = buffer.GetValues())
-                    c_api.TF_StringEncode(src, (UIntPtr)length, (sbyte*)(tensor + sizeof(Int64)), size, status);
+                    c_api.TF_StringEncode(src, (ulong)length, (byte*)(tensor + sizeof(Int64)), size, status.Handle);
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:772]: @codemzs  Why the pointer needs to +8 here

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs [hunk 4]
```diff
@@ -430,7 +431,7 @@ private protected override void SaveModel(ModelSaveContext ctx)
                 var buffer = Session.graph.ToGraphDef(status);
                 ctx.SaveBinaryStream("TFModel", w =>
                 {
-                    w.WriteByteArray(buffer.MemoryBlock.ToArray());
+                    w.WriteByteArray(buffer.DangerousMemoryBlock.ToArray());
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:434]: @Oceania2018 
Can you confirm the replacement of `MemoryBlock` here is correct?

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs [hunk 5]
```diff
@@ -1256,7 +1256,7 @@ private void VariableSummaries(RefVariable var)
             Tensor logits = null;
             tf_with(tf.name_scope(layerName), scope =>
             {
-                RefVariable layerWeights = null;
+                ResourceVariable layerWeights = null;
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:1259]: @Oceania2018 
Is the replacement correct here?

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs (hunk 1 above)
Comment (Oceania2018) [reply src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:841]: You can use `new Tensor(byte[][])` if shape is `shape(x, y)`

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs (hunk 3 above)
Comment (Oceania2018) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:772]: Tensorflow uses first 8 bytes to save string length.

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs (hunk 4 above)
Comment (Oceania2018) [reply src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:434]: Correct.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs (hunk 5 above)
Comment (Oceania2018) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:1259]: Should use `IVariableV1` interface.

File: build/Dependencies.props [hunk 6]
```diff
@@ -22,9 +22,9 @@
     <SystemDrawingCommonPackageVersion>4.5.0</SystemDrawingCommonPackageVersion>
     <SystemIOFileSystemAccessControl>4.5.0</SystemIOFileSystemAccessControl>
     <SystemSecurityPrincipalWindows>4.5.0</SystemSecurityPrincipalWindows>
-    <TensorFlowVersion>1.14.0</TensorFlowVersion>
+    <TensorFlowVersion>2.3</TensorFlowVersion>
```
Comment (eerhardt) [inline build/Dependencies.props:25]: This should be `2.3.0`. Always use a full version, so there is no ambiguity.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs [hunk 7]
```diff
@@ -752,15 +752,15 @@ private void CheckTrainingParameters(Options options)
...
             return (jpegData, resizedImage);
         }
 
         private static Tensor EncodeByteAsString(VBuffer<byte> buffer)
         {
             int length = buffer.Length;
-            var size = c_api.TF_StringEncodedSize((UIntPtr)length);
-            var handle = c_api.TF_AllocateTensor(TF_DataType.TF_STRING, IntPtr.Zero, 0, (UIntPtr)((ulong)size + 8));
+            var size = c_api.TF_StringEncodedSize((ulong)length);
+            var handle = c_api.TF_AllocateTensor(TF_DataType.TF_STRING, new long[0], 0, ((ulong)size + 8));
```
Comment (eerhardt) [inline src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:763]: Do you have to create a new, empty array here every time? Instead, would it be possible to use `Array.Empty<long>()`, so it doesn't allocate and throw away such a small object?

Comment (LittleLittleCloud) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:763]: Good point

Comment (LittleLittleCloud) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:763]: What's the difference between `Array.Empty<long>()`, `new long[0]` and `(long*)(IntPtr.Zero)`, especially why using `Array.Empty<long>()` can avoid extra mem allocate.

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs (hunk 4 above)
Comment (sharwell) [reply src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:434]: I don't think this is correct. The code needs a `using` statement around `buffer` to ensure it does not get collected before this call completes. Direct use of `DangerousMemoryBlock` will not prevent the GC from releasing it while it is still in use.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs [hunk 8]
```diff
@@ -1450,7 +1450,7 @@ private protected override void SaveCore(ModelSaveContext ctx)
             var buffer = _session.graph.ToGraphDef(status);
             ctx.SaveBinaryStream("TFModel", w =>
             {
-                w.WriteByteArray(buffer.MemoryBlock.ToArray());
+                w.WriteByteArray(buffer.DangerousMemoryBlock.ToArray());
```
Comment (sharwell) [inline src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:1453]: ⚠️ This use of `buffer` need to be protected. The easiest way is to change `var buffer` to `using var buffer`.

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs (hunk 7 above)
Comment (eerhardt) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:763]: `new long[0]` will always create a new managed array object. So new memory (to store the object) will always be allocated every time this method is called. And when the code is done with each object, they will be garbage collected.

`Array.Empty<long>()` on the other hand will allocate a single empty array for the entire process, and return that each time. See https://github.com/dotnet/runtime/blob/9c472545df220d7ae26329e1649f880c4ee127e8/src/libraries/System.Private.CoreLib/src/System/Array.cs#L702-L712.

`(long*)(IntPtr.Zero)` is completely different in that it isn't a managed array at all - it is a `nullptr` to a `long`.

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs (hunk 4 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:434]: # Resolved

File: src/Microsoft.ML.Vision/ImageClassificationTrainer.cs (hunk 8 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.Vision/ImageClassificationTrainer.cs:1453]: #Resolved

File: src/Microsoft.ML.TensorFlow/TensorFlowModel.cs [hunk 9]
```diff
@@ -6,6 +6,7 @@
 using Microsoft.ML.Runtime;
 using Microsoft.ML.TensorFlow;
 using Tensorflow;
+using static Tensorflow.Binding;
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorFlowModel.cs:9]: Is this change necessary?

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs [hunk 10]
```diff
@@ -831,13 +835,13 @@ private Tensor CastDataAndReturnAsTensor(T[] data)
                     return new Tensor((double[])(object)data, _dims, TF_DataType.TF_DOUBLE);
                 else if (typeof(T) == typeof(ReadOnlyMemory<char>))
                 {
-                    byte[][] bytes = new byte[_vBuffer.Length][];
-                    for (int i = 0; i < bytes.Length; i++)
+                    string[] strings = new string[_vBuffer.Length];
+                    for (int i = 0; i < strings.Length; i++)
                     {
-                        bytes[i] = Encoding.UTF8.GetBytes(((ReadOnlyMemory<char>)(object)data[i]).ToArray());
+                        strings[i] = new string(((ReadOnlyMemory<char>)(object)data[i]).ToArray());
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorflowTransform.cs:841]: Just call `data[i].ToString()` instead. The advantage is that if the `ROM<char>` is backed by a full string, there is no need to allocate a new string object. 

[https://github.com/dotnet/runtime/blob/c5b6881e9b0e0826029185cd5a1569f80f6b9aba/src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs#L148](https://github.com/dotnet/runtime/blob/c5b6881e9b0e0826029185cd5a1569f80f6b9aba/src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs#L148)

The way you have it here, you will get an extra `char[]` object created, and an extra `string` created.

File: src/Microsoft.ML.Vision/DnnRetrainTransform.cs [hunk 11]
```diff
@@ -1072,13 +1073,13 @@ private Tensor CastDataAndReturnAsTensor(T[] data)
                     return new Tensor((double[])(object)data, _dims, TF_DataType.TF_DOUBLE);
                 else if (typeof(T) == typeof(ReadOnlyMemory<char>))
                 {
-                    byte[][] bytes = new byte[_bufferedData.Length][];
-                    for (int i = 0; i < bytes.Length; i++)
+                    string[] strings = new string[data.Length];
+                    for (int i = 0; i < strings.Length; i++)
                     {
-                        bytes[i] = Encoding.UTF8.GetBytes(((ReadOnlyMemory<char>)(object)data[i]).ToArray());
+                        strings[i] = new string(((ReadOnlyMemory<char>)(object)data[i]).ToArray());
```
Comment (eerhardt) [inline src/Microsoft.ML.Vision/DnnRetrainTransform.cs:1079]: Same comment as elsewhere - `data[i].ToString()`.

File: src/Microsoft.ML.Vision/DnnRetrainTransform.cs [hunk 12]
```diff
@@ -1155,13 +1156,13 @@ private Tensor CastDataAndReturnAsTensor(T[] data)
                     return new Tensor((double[])(object)data, _dims, TF_DataType.TF_DOUBLE);
                 else if (typeof(T) == typeof(ReadOnlyMemory<char>))
                 {
-                    byte[][] bytes = new byte[_vBuffer.Length][];
-                    for (int i = 0; i < bytes.Length; i++)
+                    string[] strings = new string[_vBuffer.Length];
+                    for (int i = 0; i < strings.Length; i++)
                     {
-                        bytes[i] = Encoding.UTF8.GetBytes(((ReadOnlyMemory<char>)(object)data[i]).ToArray());
+                        strings[i] = new string(((ReadOnlyMemory<char>)(object)data[i]).ToArray());
```
Comment (eerhardt) [inline src/Microsoft.ML.Vision/DnnRetrainTransform.cs:1162]: Maybe it is time to refactor this into a helper method so it isn't duplicated 3 times?

File: test/Microsoft.ML.TestFramework/Attributes/TensorflowFactAttribute.cs [hunk 13]
```diff
@@ -12,14 +13,17 @@ namespace Microsoft.ML.TestFramework.Attributes
...
         {
         }
 
         /// <inheritdoc />
         protected override bool IsEnvironmentSupported()
         {
-            return Environment.Is64BitProcess;
+            return Environment.Is64BitProcess &&
+                   ( RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ||
+                     AttributeHelpers.CheckLibcVersionGreaterThanMinimum(new Version(2, 23)));
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/Attributes/TensorflowFactAttribute.cs:25]: I know this is the libc version that is the minimum Onnx supported version, but is it the same for TensorFlow?

Comment (LittleLittleCloud) [reply test/Microsoft.ML.TestFramework/Attributes/TensorflowFactAttribute.cs:25]: Yeah, it's also the same case for TF.

File: src/Microsoft.ML.TensorFlow/TensorFlowModel.cs (hunk 9 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.TensorFlow/TensorFlowModel.cs:9]: Nope, removed

File: src/Microsoft.ML.Vision/DnnRetrainTransform.cs (hunk 12 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.Vision/DnnRetrainTransform.cs:1162]: Sure

File: test/Microsoft.ML.TestFramework/Attributes/TensorflowFactAttribute.cs (hunk 13 above)
Comment (LittleLittleCloud) [reply test/Microsoft.ML.TestFramework/Attributes/TensorflowFactAttribute.cs:25]: Strictly speaking, not exactly the same, but whatever applies to disable Onnx fact also applys to tf fact.

**Review by harishsk — APPROVED** [review]
The changes look good in general. But a few of the tests are still failing. Please look into those.

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs [hunk 14]
```diff
@@ -410,6 +412,46 @@ internal static bool IsTypeSupported(TF_DataType tfoutput)
...
+            else if (typeof(T) == typeof(float))
+                return new Tensor((float[])(object)data, dims, TF_DataType.TF_FLOAT);
+            else if (typeof(T) == typeof(double))
+                return new Tensor((double[])(object)data, dims, TF_DataType.TF_DOUBLE);
+            else if (typeof(T) == typeof(ReadOnlyMemory<char>))
+            {
+                string[] strings = new string[data.Length];
+                for (int i = 0; i < strings.Length; i++)
+                {
+                    strings[i] = ((ReadOnlyMemory<char>)(object)data[i]).ToString();
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:446]: (nit) do you need the casts here? Why can't this just be
```suggestion
                    strings[i] = data[i].ToString();
```

Comment (LittleLittleCloud) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs]: # Resolved

File: src/Microsoft.ML.TensorFlow/TensorflowUtils.cs [hunk 15]
```diff
@@ -439,15 +439,7 @@ internal static Tensor CastDataAndReturnAsTensor<T>(T[] data, TensorShape tfShap
...
-            {
-                string[] strings = new string[data.Length];
-                for (int i = 0; i < strings.Length; i++)
-                {
-                    strings[i] = ((ReadOnlyMemory<char>)(object)data[i]).ToString();
-                }
-
-                return new Tensor(strings);
-            }
+                return new Tensor(data.Select(x => x.ToString()).ToArray());
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:442]: Instead of using Linq here, I would the loop and `new string[data.Length]` code that you had before. It will achieve better performance since you will only allocate a single array.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:442]: I'm not quite follow here, the size of `data` is fixed, so where will the second memory allocation be when calling `ToArray`

Comment (LittleLittleCloud) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:442]: Or is that `Select` where the n memory allocation will be

Comment (eerhardt) [reply src/Microsoft.ML.TensorFlow/TensorflowUtils.cs:442]: Correct - the `Select` will cause an allocation:

https://github.com/dotnet/runtime/blob/4486164506a56200e66cf4cd750fde6b54ab5abd/src/libraries/System.Linq/src/System/Linq/Select.cs#L37
